### PR TITLE
Start Ryuk lazily when in the reusable mode

### PR DESCRIPTION
**Pull request by bsideup**

### CODE REVIEW

File: core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+class JVMHookResourceReaper extends ResourceReaper {
+
+    JVMHookResourceReaper() {
+        setHook();
+    }
+
+    @Override
+    public synchronized void performCleanup() {
+        super.performCleanup();
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java:22]: Why do we first have to call the super method (which will call)
```java
  registeredContainers.forEach(this::removeContainer);
  registeredNetworks.forEach(this::removeNetwork);
  registeredImages.forEach(this::removeImage);
```

and then call prune afterward?

**Review by kiview — COMMENTED** [review]
The changes around lazy Ryuk for reusable mode LGTM. My questions and comments are probably mostly from a missing context of past workarounds.

File: core/src/main/java/org/testcontainers/DockerClientFactory.java [hunk 2]
```diff
@@ -174,17 +175,16 @@ public String getRemoteDockerUnixSocketPath() {
...
-            return dockerClient;
-        }
-
         // fail-fast if checks have failed previously
         if (cachedClientFailure != null) {
             log.debug("There is a cached checks failure - throwing", cachedClientFailure);
             throw cachedClientFailure;
         }
 
+        if (dockerClient != null) {
```
[the remark this answers is not in the document]
Comment (bsideup) [reply core/src/main/java/org/testcontainers/DockerClientFactory.java:184]: I think it was an artifact, really. Probably before we added `cachedClientFailure`

If checks fail, we would cache them. If ResourceReaper fails, we cache it too.

So there should be absolutely no harm in storing client earlier, and prevent us from having recursive issues

File: core/src/main/java/org/testcontainers/DockerClientFactory.java [hunk 3]
```diff
@@ -198,6 +198,7 @@ public void close() {
                 throw new IllegalStateException("You should never close the global DockerClient!");
             }
         };
+        dockerClient = client;
```
[the remark this answers is not in the document]
Comment (bsideup) [reply core/src/main/java/org/testcontainers/DockerClientFactory.java:201]: Actually, the only reason why I kept the local variable is to keep the changeset smaller :D

Happy to remove it, now that the main change is confirmed 👍

File: core/src/main/java/org/testcontainers/DockerClientFactory.java [hunk 4]
```diff
@@ -210,24 +211,15 @@ public void close() {
...
-            } catch (RuntimeException e) {
-                cachedClientFailure = e;
-                throw e;
-            }
-            log.info("Ryuk started - will monitor and terminate Testcontainers containers on JVM exit");
-        } else {
-            log.debug("Ryuk is disabled");
-            ryukContainerId = null;
-            // best-efforts cleanup at JVM shutdown, without using the Ryuk container
+        try {
```
[the remark this answers is not in the document]
Comment (bsideup) [reply core/src/main/java/org/testcontainers/DockerClientFactory.java:213]: Good point - added an explicit `ResourceReaper#init` 👍

File: core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java (hunk 1 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java:23]: `forEach` methods will cleanup resources explicitly registered with `registerContainerForCleanup`, while `DEATH_NOTE`-based ones will be cleaning by labels

File: core/src/main/java/org/testcontainers/DockerClientFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kiview) [reply core/src/main/java/org/testcontainers/DockerClientFactory.java:213]: Much clearer, thanks!

File: core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java (hunk 1 above)
Comment (kiview) [reply core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java:23]: But the set of resources each approach would remove, is very likely to overlap, or?

**Review by kiview — APPROVED** [review]
LGTM, and besides the performance improvements IMO also an improvement in clarity 👍

File: core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java (hunk 1 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java:23]: it is actually the other way around - very unlikely.

We don't register containers/networks/images by hands and rely on the filter-based approach. I deprecated the methods, so we can remove them in the future, but for now we have to keep both.

Comment (kiview) [reply core/src/main/java/org/testcontainers/utility/JVMHookResourceReaper.java:23]: That makes much more sense then, thanks for the explanation.

**Pull request merged**